pkg/cli/test.go

```diff
@@ -0,0 +1,323 @@
+					}
+				}()
+
+				tctx, span := otel.Tracer("wolfictl").Start(ctx, "test")
+				defer span.End()
+				ctx = tctx
+			}
+
+			if cfg.jobs == 0 {
+				cfg.jobs = runtime.GOMAXPROCS(0)
+			}
+
+			if cfg.pipelineDir == "" {
+				cfg.pipelineDir = filepath.Join(cfg.dir, "pipelines")
+			}
+			if cfg.outDir == "" {
+				cfg.outDir = filepath.Join(cfg.dir, "packages")
+			}
+
+			if cfg.cacheDir != "" {
+				if err := os.MkdirAll(cfg.cacheDir, os.ModePerm); err != nil {
+					return fmt.Errorf("creating cache directory: %w", err)
+				}
+			}
+
+			return testAll(ctx, &cfg, args)
+		},
+	}
+
+	cmd.Flags().StringVarP(&cfg.dir, "dir", "d", ".", "directory to search for melange configs")
+	cmd.Flags().StringVar(&cfg.pipelineDir, "pipeline-dir", "./pipelines", "directory used to extend defined built-in pipelines")
+	cmd.Flags().StringVar(&cfg.runner, "runner", "docker", "which runner to use to enable running commands, default is based on your platform.")
+	cmd.Flags().StringSliceVar(&cfg.archs, "arch", []string{"x86_64", "aarch64"}, "arch of package to build")
+	cmd.Flags().StringSliceVarP(&cfg.extraKeys, "keyring-append", "k", []string{"https://packages.wolfi.dev/os/wolfi-signing.rsa.pub"}, "path to extra keys to include in the build environment keyring")
+	cmd.Flags().StringSliceVarP(&cfg.extraRepos, "repository-append", "r", []string{"https://packages.wolfi.dev/os"}, "path to extra repositories to include in the build environment")
+	cmd.Flags().StringSliceVar(&cfg.extraPackages, "test-package-append", []string{"wolfi-base"}, "extra packages to install for each of the test environments")
+	cmd.Flags().StringVar(&cfg.cacheDir, "cache-dir", "./melange-cache/", "directory used for cached inputs")
+	cmd.Flags().StringVar(&cfg.cacheSource, "cache-source", "", "directory or bucket used for preloading the cache")
+	cmd.Flags().StringVar(&cfg.dst, "destination-repository", "", "repo where packages will eventually be uploaded, used to skip existing packages (currently only supports http)")
+	cmd.Flags().BoolVar(&cfg.debug, "debug", true, "enable test debug logging")
+
+	cmd.Flags().IntVarP(&cfg.jobs, "jobs", "j", 0, "number of jobs to run concurrently (default is GOMAXPROCS)")
+	cmd.Flags().StringVar(&traceFile, "trace", "", "where to write trace output")
+
+	return cmd
+}
+
+type testConfig struct {
+	archs         []string
+	extraKeys     []string
+	extraRepos    []string
+	extraPackages []string
+
+	outDir      string // used for keeping logs consistent with build
+	dir         string
+	dst         string
+	pipelineDir string
+	runner      string
+	debug       bool
+
+	cacheSource string
+	cacheDir    string
+
+	jobs int
+}
+
+func testAll(ctx context.Context, cfg *testConfig, packages []string) error {
+	log := clog.FromContext(ctx)
+
+	pkgs, err := cfg.getPackages(ctx)
+	if err != nil {
+		return fmt.Errorf("getting packages: %w", err)
+	}
+
+	todoPkgs := make(map[string]struct{}, len(packages))
```

Comment:
nit: I would find it easier to read this, if we constructed the list of packages that we will test here up front instead of creating these two data structures (pkgs, todoPkgs) and then in the for loop below filter things out. Something like, if len(packages) > 0, then maybe use pkgs.Sub(packages...) otherwise just use the full one. Thinking being that I'd expect the common case to be 1 or few packages.
https://github.com/wolfi-dev/wolfictl/blob/v0.15.18/pkg/dag/packages.go#L349